File: leetcode_python/239.Sliding_Window_Maximum.py
#!/usr/bin/env python3
# coding: utf-8

# Time complexity: O()
# Space complexity: O()

# https://leetcode.com/problems/sliding-window-maximum/

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        if not nums or k == 0:
            return []
        length = len(nums)
        if length <= k:
            return [max(nums)]
        queue = collections.deque([])
        res = []
        for i in range(length):
            if queue and queue[0] < i - k + 1:
                queue.popleft()
            while queue and nums[queue[-1]] < nums[i]:
                queue.pop()
            queue.append(i)
            if i >= k - 1:
                res.append(nums[queue[0]])

        return res

File: leetcode_go/239.Sliding_Window_Maximum.go
// https://leetcode.com/problems/sliding-window-maximum/

func maxSlidingWindow(nums []int, k int) []int {
    deque := make([]int, 0, k)
    res := make([]int, 0, len(nums)-k+1)
    
    for i, value := range nums {
        if len(deque) > 0 && deque[0] <= i - k {
            deque = deque[1:]
        }
        
        for len(deque) > 0 && nums[deque[len(deque)-1]] <= value {
            deque = deque[:len(deque)-1]
        }
        
        deque = append(deque, i)
        
        if i >= k - 1 {
            res = append(res, nums[deque[0]])
        } 
    }
    
    return res
}